### PR TITLE
rag: annotate with chunk scores, make disambiguator gentle on its task

### PR DESCRIPTION
**Pull request by mamuqsit**
## Introduction
System prompt for disambiguation agent is all-encompassing - it does not leave much room for user queries to be rightfully categorized as IDKs. One extreme example is of when the user enquires "What is elephantiasis?" and the bot responds with a disambiguation based on chunks it retrieved (it retrieved chunks with sim=0.02).

## Changes
- Annotate KB chunks with `score` before passing to LLMs - agent gets to reason whether these chunks are relevant (rather than us defining 'magic numbers' for cosine sim thresholds). https://github.com/A4i-tech/byoeb/pull/136/commits/a7784bd427a358be860ad3d16d5da3951bd6caf2
- Remove chunks that do not contain any alphanumeric symbols https://github.com/A4i-tech/byoeb/pull/136/commits/71960f2fcd991e59f8ba8fd6e9fc6387c61ac0e6
- Remove chunks with low cosine similarity scores (search=dense, t<.35) https://github.com/A4i-tech/byoeb/pull/136/changes/5d7936f0f464634127c534a489d306135e393351
- Gate disambiguation behind feature flag `ASHABOT_FEATURE_FLAGS=query_disambiguation`. This feature is otherwise rolled only to test users. https://github.com/A4i-tech/byoeb/pull/136/changes/5d7513659a6d80fd1f7cba5fdfd4e1a1ef455972

## Testing
1. Start ASHABot
2. Expose port 8000 - `ngrok http 8000`
3. Run code on https://colab.research.google.com:
   ```py
   # !pip install fastmcp==2.14.0
   from fastmcp import Client

   URL = "https://750b-49-207-57-7.ngrok-free.app/mcp?phone_number=91xxxx"
   async def ask(q):
       async with Client(URL) as client:
           resp = await client.call_tool("asha_chat", {"message": q})
           return resp.data.text

   print(await ask("benefits of vitamin k"))
   ```

### CODE REVIEW

**Review by a4i-architect — COMMENTED** [review]
### High-Level Review Summary

The modifications mainly intend to refactor the way KB chunks are converted into string summaries for prompts, replacing inline string joining with a dedicated `_chunks_to_kb_topics()` helper method for reusability and cleaner code. In the `chat_app/bot_config.json` update, the prompt was clarified to include essential guidance about relevance checks and decision logic, improving user question clarity handling. 

In `generate.py`, the key change is the introduction of `_chunks_to_kb_topics()`, used in two places for converting chunks into textual summaries with a consistent format.

---

### Issues & Recommendations

1. **Type Safety & Implementation Bugs**
   - `_chunks_to_kb_topics()` claims `chunks: Iterable[Chunk]`, but it's called with generator expressions. While technically acceptable, enforce type consistency or convert to list if repeated iteration is needed.
   - In `_chunks_to_kb_topics()`, the generator expression `for chunk in chunks` implies `chunks` is already iterable. But in `needs_clarification()` and elsewhere, it's invoked with generator expressions. Prefer explicitly converting to list inside the method for clarity.

2. **Code Duplication & Naming**
   - The method `_chunks_to_kb_topics()` duplicates functionality of previously inline join logic; ensure consistent format by defining and reusing it everywhere.
   
3. **Logical & Design Aspects**
   - The modified prompt in `system_prompt` adds a crucial step to check for the relevance of KB chunks before asking clarification. This is good, but it's a complex change that should be clearly documented for future maintainers, emphasizing the importance of relevance filtering.
   - The implementation of `needs_clarification()` now relies on `_chunks_to_kb_topics()`, which outputs a formatted string, but the method should ideally return structured data for clarity and testability, not just a string.

4. **Potential Over-Engineering**
   - The `_chunks_to_kb_topics()` method is simple, but its use in multiple places suggests a minor enhancement rather than a significant pattern shift. Keep an eye on whether this abstraction adds value or complicates debugging.

5. **Testing & Validation**
   - No tests are shown; ensure that the new formatting function is covered in unit tests, especially with diverse chunk inputs and edge cases.

---

### Overall Rating: 75/100
- Strengths: Clear intention to improve maintainability and prompt clarity, improved prompt instructions.
- Improvements Needed: Better type safety, consistency, more explicit handling of chunk data, supplementary tests, and documentation for the relevance check logic.

**Final note:** These refactorings are beneficial but require careful testing to confirm behavior aligns with intended clarification logic and prompt guidance, especially considering the importance of relevance filtering in the new prompt.

**Review by a4i-architect — COMMENTED** [review]
### Code Review Summary

**Key Changes & Observations:**

1. **Chat App JSON (`bot_config.json`):**
   - **Major improvement**: The `system_prompt` for clarification now emphasizes prior validation of relevance before requesting clarification, adding a critical step for avoiding unnecessary prompts.
   - **Clarity & Context**: The revised prompt provides clearer, more directive instructions, including explicit rules and explicit XML formatting instructions.
   - **Potential issue**: The `response_translate` section remains unchanged, assuming proper translation logic elsewhere. Verify if this translation should be dynamic or static.

2. **Constants (`feature_enums.py`):**
   - Added a new feature flag `QUERY_DISAMBIGUATION` which is sensible for controlling clarification behavior.
   - No apparent issues; maintains good extensibility.

3. **Chat Handling (`generate.py`):**
   - Introduced `_chunks_to_kb_topics()` for formatting chunks into an XML string; this promotes code reuse.
   - Minor typo: in the `_chunks_to_kb_topics()` function, the parameter should be an `Iterable[Chunk]`, but in the call inside `parse_response_xml`, the arguments are generator expressions. Make sure the generator is converted to a list or properly iterated.
   - The `needs_clarification()` method now directly converts chunks to the formatted string; this improves clarity.
   - The main logic now checks if chunks contain relevant info before prompting clarification, aligning with the new system prompt instructions.

**Issues & Recommendations:**

- **Type consistency & generator handling**: Ensure generator expressions are converted to lists where necessary, especially in `_chunks_to_kb_topics()`.
- **Code duplication**: The `_chunks_to_kb_topics()` method is well abstracted, but verify if similar formatting is needed elsewhere to avoid redundant code.
- **Logical considerations**:
  - The new prompt logic strongly depends on the `KB_TOPICS` relevance check. Make sure the `KB_TOPICS` relevance is computed correctly upstream.
  - In `handle_message_generate_workflow`, the current check for `is_idk()` combined with a feature flag is good, but confirm if fallback to clarification should also consider the actual content relevance.

**Design & Complexity:**

- The improvements in prompt instructions reduce ambiguity and improve chat clarity.
- Usage of feature flags is a good pattern, enhancing configurability.
- Overall, logic is reasonable; avoid overcomplicating the `needs_clarification()` flow.

**Score:** 82/100

**Final notes**:  
- Double-check the actual relevance filtering for `KB_TOPICS`.  
- Perform unit tests on `_chunks_to_kb_topics()` with mock Chunk data.  
- Confirm translation logic for `response_translate`.

**References**:  
- Effective prompt engineering, [OpenAI guide](https://openai.com/research/instruction-following).  
- Code reuse best practices, [Martin Fowler](https://martinfowler.com/bliki/ExtractMethod.html).

**Review by N1kunj1998 — APPROVED** [review]
Tested on local and working.

File: byoeb-v1/byoeb/byoeb/services/chat/message_handlers/user_flow_handlers/generate.py [hunk 1]
```diff
@@ -522,6 +522,12 @@ def __create_expert_verification_message(
             incoming_timestamp=message.incoming_timestamp,
         )
         return new_expert_verification_message
+
+    def _chunks_to_kb_topics(self, chunks: Iterable[Chunk]) -> str:
+        return "\n".join(
```
Comment (Soumabha) [inline byoeb-v1/byoeb/byoeb/services/chat/message_handlers/user_flow_handlers/generate.py]: Lets gate keep any empty chunk.text from being joined and explicitly remove chunks with low / less score (guarded by a configurable threshold)

File: byoeb-v1/byoeb/byoeb/chat_app/bot_config.json [hunk 2]
```diff
@@ -42,7 +42,7 @@
...
...the KB in simple, conversational language\n- Make it easy to respond with just a number, word, or \"yes\"\n- Keep it SHORT and scannable\n- Use everyday language\n\nExamples of what NOT to do:\n- \u274c \"Are you asking about a specific program or just in general?\"\n- \u274c \"Could you clarify which benefit you mean?\"\n- \u274c \"Are you referring to scheme A or scheme B?\"\n\nExamples of what TO do:\n- \u2713 \"I found a few things about money for girl child birth:\n1. Cash benefit under JSY program\n2. Free services at hospitals\n3. Delayed marriage incentive\n\nWhich one do you want to know about?\"\n- \u2713 \"There are different benefits available:\n- Direct cash payment (\u20b9X)\n- Free hospital services\n- Nutrition support\n\nTell me which interests you?\"\n- \u2713 \"I see info on:\n1. Government cash schemes\n2. Hospital benefits\nWant details on both or just one?\"\n\nDo NOT answer the question yourself - just present the options.\n\nWhen presenting options, output in XML format as shown below.\n\n<o>\n<clarification_en>Present available options conversationally in English</clarification_en>\n<clarification_src>Options in source language</clarification_src>\n</o>\n",
```
Comment (Soumabha) [inline byoeb-v1/byoeb/byoeb/chat_app/bot_config.json]: The score interpretation shoould also be explained to the LLM, the prompt needs to specify score meaning, range, "high" vs "low" details in addition to the instruction for validating relevance.

File: byoeb-v1/byoeb/byoeb/services/chat/message_handlers/user_flow_handlers/generate.py [hunk 3]
```diff
@@ -762,7 +768,7 @@ async def handle_message_generate_workflow(
                             response_en, response_source, tokens = response_en2, response_source2, tokens2
                             break
 
-                if utils.is_idk(response_en):
+                if utils.is_idk(response_en) and (FeatureFlag.QUERY_DISAMBIGUATION in feature_flags or message.user.test_user):
```
Comment (Soumabha) [inline byoeb-v1/byoeb/byoeb/services/chat/message_handlers/user_flow_handlers/generate.py:796]: Pls see if we need to import FeatureFlag

File: byoeb-v1/byoeb/byoeb/services/chat/message_handlers/user_flow_handlers/generate.py [hunk 4]
```diff
@@ -522,6 +522,12 @@ def __create_expert_verification_message(
             incoming_timestamp=message.incoming_timestamp,
         )
         return new_expert_verification_message
+
+    def _chunks_to_kb_topics(self, chunks: Iterable[Chunk]) -> str:
```
Comment (Soumabha) [inline byoeb-v1/byoeb/byoeb/services/chat/message_handlers/user_flow_handlers/generate.py:543]: Pls add unit tests to evaluate the formatting in different scenarios.
1. Test with lower score than threshold
2. Test with empty / None chunk text
3. Test with empty list of chunks
4. Test with normal case.

File: byoeb-v1/byoeb/byoeb/services/chat/message_handlers/user_flow_handlers/generate.py (hunk 3 above)
Comment (mamuqsit) [reply byoeb-v1/byoeb/byoeb/services/chat/message_handlers/user_flow_handlers/generate.py:796]: Yeah it is imported in the file - embedding cache flow uses the same logic couple of lines above

**Review by a4i-architect — COMMENTED** [review]
### High-Level Summary

The code review reveals thoughtful enhancements in retrieval filtering, clarification prompts, and feature flags. Key improvements include:

- Introduction of `retrieval.similarity_threshold` configuration for refined chunk filtering.
- Addition of `QUERY_DISAMBIGUATION` feature flag for user query disambiguation.
- Refined retrieval chunk filtering with `filter_retrieved_chunks` and transformations to KB topic formatting.
- Updated clarification prompts, especially the system prompt for better contextual understanding and user guidance.
- Unit tests cover chunk filtering logic thoroughly.

### Issues & Recommendations

1. **Functional & Typing Issues**:
   - `filter_retrieved_chunks` method:
     - The generator expression should convert to a list before use in `_chunks_to_kb_topics`. Passing a generator directly causes iteration issues and potential bugs.
   - Typings:
     - Explicitly annotate `filter_retrieved_chunks` method as returning `List[Chunk]`.
     - In `_chunks_to_kb_topics`, accept `Iterable[Chunk]` for clarity.
   - `re.sub(r'\W+', '', chunk.text)` filtering:
     - Ensure `chunk.text` is always a string or handle None safely to avoid runtime errors.

2. **Code Duplication & Clarity**:
   - The pattern of filtering chunks by source (e.g., `"KB Updated"`) repeats. Consider DRY principles or helper functions for tags.
   - The new `filter_retrieved_chunks` method is well-conceived but needs proper integration to avoid partial filtering.

3. **Design & Complexity**:
   - Clarification prompt content is highly verbose; consider externalizing large prompt text into separate template files for maintainability.
   - Feature flag checking (`FeatureFlag.QUERY_DISAMBIGUATION in feature_flags or message.user.test_user`) appears sound but verify feature flag population and consistency.

4. **Testing & Coverage**:
   - The test rigorously covers filtering behavior but should also test edge cases:
     - Null or missing `chunk.text`
     - Large datasets for performance
     - Cases where no chunks meet threshold
   
### Score: 80/100 (Acceptable with improvements)

### Final Note
Refine typing, ensure generator usage correctness, and modularize prompt contents. These improvements will enhance code robustness, clarity, and maintainability.

File: byoeb-v1/byoeb/byoeb/services/chat/message_handlers/user_flow_handlers/generate.py (hunk 1 above)
Comment (mamuqsit) [reply byoeb-v1/byoeb/byoeb/services/chat/message_handlers/user_flow_handlers/generate.py]: Addressed in https://github.com/A4i-tech/byoeb/pull/136/changes/18a059c9415e058235d2f48490324a2114616fe9#diff-1febc972bb176d4dc42e902fd019b737d5d845548d866d0aadfe50cf311d9f05R526-R531

File: byoeb-v1/byoeb/byoeb/services/chat/message_handlers/user_flow_handlers/generate.py (hunk 4 above)
Comment (mamuqsit) [reply byoeb-v1/byoeb/byoeb/services/chat/message_handlers/user_flow_handlers/generate.py:543]: Addressed in https://github.com/A4i-tech/byoeb/pull/136/changes/e2076000e26b4433e5f331985b20856b6e24ec44

File: byoeb-v1/byoeb/byoeb/chat_app/bot_config.json (hunk 2 above)
Comment (mamuqsit) [reply byoeb-v1/byoeb/byoeb/chat_app/bot_config.json]: Addressed in https://github.com/A4i-tech/byoeb/pull/136/changes/54da6ec88207521a467e17ce50b5fdb94130491b

**Review by a4i-architect — COMMENTED** [review]
### High-Level Summary

The overall codebase has been enhanced to improve chunk filtering, retrieval, and clarification prompt generation. Key changes introduce a dedicated `filter_retrieved_chunks` method with improved filtering logic, a method to convert chunks to text blocks for prompts, and a new feature flag for query disambiguation. The JSON config has been updated to include `retrieval.similarity_threshold` and `QUERY_DISAMBIGUATION` flag. The unit tests sufficiently cover chunk filtering behavior.

### Key Issues & Recommendations

1. **Type Annotations & Method Scope**:
   - `filter_retrieved_chunks` and `_chunks_to_kb_topics` are instance methods used internally but lack explicit class context in the snippets. Ensure these methods are properly indented and within the class scope.
   - The filter method takes an `Iterable[Chunk]` but returns a generator. It might be clearer to explicitly annotate return as `Iterator[Chunk]` for clarity.

2. **Filtering Logic**:
   - The filtering excludes chunks with empty or whitespace-only text, which is good.
   - The check `chunk.text and len(re.sub(r'\W+', '', chunk.text)) > 0` is acceptable but consider pre-caching cleaned text or simplifying if performance is critical.

3. **Type Safety & Consistency**:
   - In the `agenerate_answer` method, the `_chunks_to_kb_topics` is called with a generator. Ensure the method can handle this; explicitly convert to `list()` if needed.
   - The `Chunk` class references suggest it has attributes: `similarity`, `text`, `chunk_id`. Confirm these are present.

4. **Code Duplication & Clarity**:
   - The pattern of filtering chunks and converting to a specific string format repeats. Consider abstracting this into a reusable method.

5. **Configuration**:
   - Introduction of `retrieval.similarity_threshold` is good. But ensure its defaults and validation are handled at startup.
   - The `QUERY_DISAMBIGUATION` feature flag is introduced; verify this flag's usage logic is consistent.

6. **Testing Coverage**:
   - The unit tests are well-designed for filtering logic. However, test the full pipeline: from retrieval, filtering, prompt generation, and response handling to ensure integration consistency.
   - Add tests for `filter_retrieved_chunks` edge cases and invalid data (nulls, malformed chunks).

### Design & Complexity
- The filtering function is straightforward, but the prompt generation includes nuanced instructions. Be cautious about maintaining a simple, maintainable chain of methods; excessive inline logic might obscure intent.
- The overall pattern of filtering chunks based on a relevance score and content length is sound but validate that scoring thresholds are correctly applied across the codebase.

### Score & Acceptance
- Core filtering logic and configuration updates are beneficial, but implementation details (method placements, type annotations) could be clearer.
- Given the extensive test coverage for chunk filtering, the fix is solid with minor refactoring for clarity.

**Score: 82/100**

**Recommendations**:
- Clarify method placements and improve type annotations.
- Add edge-case tests for chunk filtering and prompt generation.
- Document the use and expected behavior of new config and feature flags.

Overall, a solid incremental improvement suitable for merging post minor refinements.

**Comment by mamuqsit** [comment]
Hm, hit a deadend - similarity score closer to 0 does not necessarily mean "bad chunk". This is true for vector search which is configured to use cosine similarity. However, hybrid search returns valid chunks but often with low scores.

<details>
<summary>Hybrid search result for query="what is the antara injection"</summary>

```json
{
    "value": [
        {
            "@search.score": 0.03333333507180214,
            "id": "f0efb4f9-28b7-4d55-bc6a-2c5383464c39",
            "text": "Frequently Asked Questions by ASHAs \r\nabout \r\nDMP A: Depot Medroxy Progesterone Acetate\r\nQ1. What is Antara (DMPA) injection?\r\nInjectable contraceptive or DMPA or Antara, \r\nas it is known in India, is a reversible \r\ncontraception method. Any women of 15-45 \r\nyears of age can use this. It is available free \r\nof cost in government run health centre and \r\nhospital.\r\nQ2. How do injectables work?\r\nInjectables work by preventing the release of the egg in the woman. If the man’s \r\nsperm has no egg to meet, the woman cannot get pregnant.\r\nAntara thickens cervical muscles to \r\nblock sperm and suppresses\r\nhormones\r\nresponsible for\r\novulation.\r\nQ3. What are failure rates of Antara (DMPA) use?\r\nDMPA has a failure rate of 0.3% when the drug is used correctly. It means it can fail \r\nin about 3 out of 1000 women.\r\nQ4. How often do women have to take the injection?\r\nThe injection has to be taken every 3 months or 13 weeks.",
            "metadata": {
                "source": "raw_documents/FAQ_Antara_edited.txt",
                "creation_timestamp": "1764677338",
                "update_timestamp": "1764677338"
            },
            "related_questions": {
                "en": [
                    "What is the age range for DMPA users?",
                    "How does Antara prevent pregnancy?",
                    "What is the DMPA failure rate?"
                ],
                "hi": [
                    "DMPA उपयोगकर्ताओं के लिए आयु सीमा क्या है?",
                    "अंतरा गर्भावस्था को कैसे रोकता है?",
                    "DMPA की विफलता दर क्या है?"
                ],
                "mr": [],
                "te": []
            }
        }
    ]
}
```
</details>

**Comment by Soumabha** [comment]
As correctly pointed out in https://github.com/A4i-tech/byoeb/pull/136#issuecomment-3883615552, since the algorithms weigh and score the retrieved chunks differently - it does not make sense to have a filter threshold across the ones which are not actually gauging similarity.

Screenshot 2026-02-11 at 6 16 28 PM

@mamuqsit Some solutions to counter the problem:
1. Apply filtering threshold differently for the algorithms, fit for cosine based similarity but can keep using the ranking (top 3) for BM25 and hybrid. We would also need to alter the system prompt to pay heed to scoring differently based on the algorithm chosen.
2. Remove filtering totally, and passing the algorithms and definition as applicable with the retrieved chunks with the basic philosophy around 'judge the relevance of the chunk with the question asked'.

**Review by a4i-architect — COMMENTED** [review]
### High-Level Summary

The code introduces enhancements to the chunk retrieval and filtering logic, notably adding `retrieval_type` annotations, filtering chunks based on similarity thresholds, and converting chunks into a topic string format for prompt construction. The JSON config is updated with a `retrieval` section for thresholds, and a new feature flag improves query disambiguation handling.

### Key Issues & Improvements

1. **Code Quality & Missing Checks:**
   - The `filter_retrieved_chunks` method assumes `chunk.metadata.additional_metadata` exists; it’s better to centralize checks to prevent potential `AttributeError`.
   - In `_chunks_to_kb_topics`, the parameter should be explicitly typed (`Iterable[Chunk]`) rather than `chunk for chunk in ...`.
   - The `test_filter_retrieved_chunks` test is a good addition, but it relies heavily on the assumed structure of `Chunk`. Ensure `Chunk` supports all used attributes.
   - The `annotate_retrieval_type()` method adds faults if `chunk.metadata` is None, but may not be called consistently elsewhere.

2. **Design & Complexity:**
   - Adding retrieval type annotation per chunk is good, but the logic for filtering and conversion (particularly `filter_retrieved_chunks`) could be made more modular and clearer.
   - The distinction and handling of different `search_type` and `retrieval_type` should be cleaner, possibly by encapsulating this logic.

3. **Type-Safety & Consistency:**
   - The code for filtering chunks based on `similarity` and `retrieval_type` should consider default values for unknown types to prevent KeyErrors (done in the test but could be safer).
   - Some functions could have more explicit typing, and the use of `Optional[str]` vs `str | None` should be consistent.

4. **Unnecessary/Redundant Code:**
   - The comment about `# cache user queries in runtime` in enum file is fine but could be clarified.
   - The configuration JSON update is straightforward, but ensure the new thresholds are well documented.

5. **Test Coverage & Robustness:**
   - The test is comprehensive, covering threshold boundary, quality filtering, and per-type thresholds.
   - Add tests for missing or malformed metadata to verify robustness.

### Overall Score: **78/100**

- **Strengths:** Clear chunk filtering enhancements, retrieval type annotation, detailed tests.
- **Improvements Needed:** Better null safety, clearer separation of retrieval logic, explicit typings, and handling of edge cases.

**Recommended action:** Refine null handling, ensure type safety, and improve modularity before acceptance.

**Comment by mamuqsit** [comment]
@Soumabha Went with the first approach https://github.com/A4i-tech/byoeb/pull/136/commits/5d7936f0f464634127c534a489d306135e393351 - threshold values are configurable per-algo (currently only configured for dense: .35).

**Review by Soumabha — APPROVED** [review]
